config/vufind/config.ini

```diff
@@ -808,10 +808,14 @@ authors         = Wikipedia
 ;                coordinate field will be displayed before the map label in the label popup.
 ; graticule: true or false. Default is false. If graticule is true a lat/long grid will be
 ;                displayed on the map.
+; basemap:   Tileserver URL for basemap. Options are osm-intl (default), stamen-toner,
+;            stamen-terrain, cartocdn-light, or cartocdn-dark. To add an alias, edit the
+;            geo_basemaps.txt file located at /vufind/config/vufind/.
```

Comment:
Is there a compelling reason to use a tab-delimited text file here? It's not the most readable of format options. Might it make sense to use an .ini, JSON or YAML format instead for greater consistency with other VuFind configuration files? If using an .ini, the section name could be the alias, and it could contain url and attribution options. A similar structure could be built with JSON or YAML if more flexibility was needed... but for this case, .ini may actually be perfectly adequate. (Using .ini would also add the benefit of being able to take advantage of VuFind's built in .ini inheritance functionality through Parent_Config, etc.).

Reply:
That's a great idea Demian. I like the idea of creating a geofeatures.ini file. What do you think of putting all of the geographic search and display configuration options into one geofeatures.ini file? That way we could pull the configuration settings out of config.ini and searches.ini and it goes into one file. Probably would be easier for long-term maintenance of the code. What do you think?

Reply:
There are a couple of things to think about here:

1.) Right now, some of the geographic features can be configured in two different places. This is either a good thing because it adds flexibility, or it's a bad thing because it requires you to configure everything twice. My gut feeling is that most of the time, users would want these settings to be uniform, and so creating a single .ini file would be an improvement... but we might want to allow the ability to configure somewhere which .ini file is read, defaulting to geofeatures.ini, so that it's still possible to create context-sensitive configurations if necessary.

2.) One disadvantage to the .ini format is that it doesn't allow a lot of nesting. You basically have sections and settings, and in a pinch, you can treat some settings as arrays, for a maximum depth of three levels. My original idea for .ini is that you would have the section name be the name of the basemap, and the settings inside specify the parameters for that map. However, having a design with arbitrary section names doesn't play well with including other types of settings, because you run into a risk of collisions (e.g. if you have a [general] section and then somebody tries to name a basemap 'general'). You could probably work around this by having a [basemaps] section, and using arbitrary key names as arrays, like:

<pre>
[Basemaps]
basemap1[url] = http://foo
basemap1[license] = foo

basemap2[url] = http://bar
basemap2[license] = bar
</pre>

...but that's a little less elegant.

Not to say that I have an opinion one way or the other about how to proceed -- but I think those are the important issues to think about when making the decision.

I'd also suggest that no matter what strategy you select, let's not move anything as part of this pull request. We'll just put the basemaps in the place where we ultimately want them to end up, and then we can take care of moving existing settings in a separate PR. That will help us focus on finishing this part before we tackle some new challenges.

Reply:
Sounds good. I would still opt for a geofeatures.ini file that we eventually migrate all the geo-related configuration settings into (see [geofeatures.ini.txt](https://github.com/vufind-org/vufind/files/1581115/geofeatures.ini.txt) for a general idea). 

One thing we could do is create a geofeatures.ini file like attached, and then check the MapSelection (for map search) and MapTab (for map display) sections for the basemap_url and basemap_attribution settings first. If those settings aren't there, then we can fall back to the Basemaps section, and if that fails, we default to the osm-intl basemap. (and for backwards compatibility we can check config.ini and searches.ini before we check geofeatures.ini for the non-basemap related settings).

But for now, I agree - we can just put the basemap settings in geofeatures.ini and look to do a separate PR where we move the rest of the configuration settings over to that ini file.

Reply:
Sounds good. Another thing we can consider is adding some logic to VuFind\Config\Upgrade so that the upgrade script is aware of the moved settings and can either relocate them or at very least warn the user about deprecations so that we can eventually eliminate some of the more complicated extra checking.